File: python/implement_queue_using_stacks.py
class MyQueue:

    def __init__(self):
        self.stack_1: list[int] = []
        self.stack_2: list[int] = []

    def push(self, x: int) -> None:
        """time complexity: O(1)"""
        self.stack_1.append(x)

    def pop(self) -> int:
        """time complexity: O(n)"""
        if len(self.stack_2) == 0:
            while len(self.stack_1) > 0:
                self.stack_2.append(self.stack_1.pop())
        return self.stack_2.pop()

    def peek(self) -> int:
        """time complexity: O(n)"""
        if len(self.stack_2) == 0:
            while len(self.stack_1) > 0:
                self.stack_2.append(self.stack_1.pop())
        return self.stack_2[-1]

    def empty(self) -> bool:
        """time complexity: O(1)"""
        return max(len(self.stack_2), len(self.stack_1)) == 0


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()
def test_MyQueue_case_1():
    obj: MyQueue = MyQueue()
    obj.push(1)
    obj.push(2)
    param_1: int = obj.peek()
    assert 1 == param_1
    param_2: int = obj.pop()
    assert 1 == param_2
    assert not obj.empty()

File: rust/src/code/implement_queue_using_stacks.rs
struct MyQueue {
    stack_1: Vec<i32>,
    stack_2: Vec<i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyQueue {
    fn new() -> Self {
        Self {
            stack_1: Vec::new(),
            stack_2: Vec::new(),
        }
    }

    // NOTE: time complexity: O(1)
    fn push(&mut self, x: i32) {
        self.stack_1.push(x);
    }

    // NOTE: time complexity: O(n)
    fn pop(&mut self) -> i32 {
        if self.stack_2.len() == 0 {
            while let Some(x) = self.stack_1.pop() {
                self.stack_2.push(x);
            }
        }
        return self.stack_2.pop().unwrap();
    }

    // NOTE: time complexity: O(n)
    fn peek(&mut self) -> i32 {
        if self.stack_2.len() == 0 {
            while let Some(x) = self.stack_1.pop() {
                self.stack_2.push(x);
            }
        }
        return *self.stack_2.last().unwrap();
    }

    // NOTE: time complexity: O(1)
    fn empty(&self) -> bool {
        self.stack_1.len() == 0 && self.stack_2.len() == 0
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * let obj = MyQueue::new();
 * obj.push(x);
 * let ret_2: i32 = obj.pop();
 * let ret_3: i32 = obj.peek();
 * let ret_4: bool = obj.empty();
 */
#[cfg(test)]
mod test {
    use super::MyQueue;

    #[test]
    fn test_my_queue_case_1() {
        let mut queue: MyQueue = MyQueue::new();
        queue.push(1);
        queue.push(2);
        assert_eq!(1, queue.peek());
        assert_eq!(1, queue.pop());
        assert_eq!(false, queue.empty());
    }
}
